### Home Work 5 Question 2
### Author: Mahmud Hasan
### Student ID: 200076270

#### Here we are importing the necessary libraries.

In [3]:
import pandas as pd
import numpy as np

#### Here we are displaying the imported dataset and showing the first few rows.

In [ ]:
f = r"C:\Users\mhasa\Google Drive\Tutorial Corner\PYTH\PyCharm Projects\PyPlay\or504_hw5"
df = pd.read_csv(f+'\places.csv', index_col=0)
df.head()

FileNotFoundError: File b'places.csv' does not exist

#### Now we shall be transforming the data in the imported data table using a base 10 log.

In [3]:
log_transformed_data = df.apply(lambda x : np.log10(x))

In [4]:
# Here we are displaying the logged data

log_transformed_data.head()

Climate  Housingcost  HealthCare     Crime  \
Abilene,TX                  2.716838     3.792392    2.374748  2.965202   
Akron,OH                    2.759668     3.910518    3.219060  2.947434   
Albany,GA                   2.670246     3.865637    2.790988  2.986772   
Albany-Schenectady-Troy,NY  2.677607     3.898067    3.155640  2.785330   
Albuquerque,NM              2.818885     3.923917    3.267875  3.171141   

                               Trans      Educ      Arts   Recreat      Econ  
Abilene,TX                  3.605413  3.440437  2.998259  3.147676  3.882695  
Akron,OH                    3.688687  3.387034  3.745387  3.420286  3.638489  
Albany,GA                   3.403292  3.408240  2.374748  2.933993  3.720159  
Albany-Schenectady-Troy,NY  3.837778  3.531351  3.667920  3.208710  3.768194  
Albuquerque,NM              3.816771  3.480869  3.652826  3.416973  3.757927

#### Now we will find out the 9 x 9 correlation matrix of the provided data.

In [5]:
correlation_matrix = log_transformed_data.corr()

In [6]:
# Diplaying the correlation matrix

correlation_matrix

Climate  Housingcost  HealthCare     Crime     Trans      Educ  \
Climate      1.000000     0.272964    0.150561  0.227751  0.021559  0.077458   
Housingcost  0.272964     1.000000    0.431935  0.139234  0.317732  0.202088   
HealthCare   0.150561     0.431935    1.000000  0.183625  0.418850  0.464764   
Crime        0.227751     0.139234    0.183625  1.000000  0.273852  0.055508   
Trans        0.021559     0.317732    0.418850  0.273852  1.000000  0.311237   
Educ         0.077458     0.202088    0.464764  0.055508  0.311237  1.000000   
Arts         0.172683     0.508501    0.678129  0.346462  0.547634  0.347899   
Recreat      0.120610     0.460696    0.254036  0.292124  0.390684  0.093001   
Econ        -0.100727     0.297058    0.054047  0.276182  0.062680  0.128858   

                 Arts   Recreat      Econ  
Climate      0.172683  0.120610 -0.100727  
Housingcost  0.508501  0.460696  0.297058  
HealthCare   0.678129  0.254036  0.054047  
Crime        0.346462  0.292124  0.276182  
Trans        0.547634  0.390684  0.062680  
Educ         0.347899  0.093001  0.128858  
Arts         1.000000  0.496519  0.134761  
Recreat      0.496519  1.000000  0.175914  
Econ         0.134761  0.175914  1.000000

covar mat

In [7]:
log_transformed_data.cov()

Climate  Housingcost  HealthCare     Crime     Trans      Educ  \
Climate      0.012892     0.003268    0.005479  0.004374  0.000386  0.000442   
Housingcost  0.003268     0.011116    0.014596  0.002483  0.005279  0.001070   
HealthCare   0.005479     0.014596    0.102728  0.009955  0.021153  0.007478   
Crime        0.004374     0.002483    0.009955  0.028611  0.007299  0.000471   
Trans        0.000386     0.005279    0.021153  0.007299  0.024829  0.002462   
Educ         0.000442     0.001070    0.007478  0.000471  0.002462  0.002520   
Arts         0.010689     0.029226    0.118484  0.031947  0.047041  0.009520   
Recreat      0.002573     0.009127    0.015299  0.009285  0.011567  0.000877   
Econ        -0.000966     0.002646    0.001463  0.003946  0.000834  0.000546   

                 Arts   Recreat      Econ  
Climate      0.010689  0.002573 -0.000966  
Housingcost  0.029226  0.009127  0.002646  
HealthCare   0.118484  0.015299  0.001463  
Crime        0.031947  0.009285  0.003946  
Trans        0.047041  0.011567  0.000834  
Educ         0.009520  0.000877  0.000546  
Arts         0.297173  0.050860  0.006206  
Recreat      0.050860  0.035308  0.002792  
Econ         0.006206  0.002792  0.007137

#### (i) Now we will be finding the 9 eigenvalues of the 9 x 9 correlation matrix.
The output array is an array of the 9 eigen values in descending order.

In [148]:
eigenvalues, eigenvectors = np.linalg.eig(correlation_matrix.values)
eigenvectors = -eigenvectors
eigenvalues.sort()
eigenvalues = eigenvalues[::-1]
eigenvalues

array([3.297793  , 1.21356187, 1.10552987, 0.90727984, 0.86062869,
       0.5621858 , 0.48382061, 0.31807215, 0.25112816])

#### (ii) Now we would be finding out the eigenvectors corresponding to the eigenvalues found in (i)

In [149]:
eigenvectors

array([[ 0.15794137,  0.06862938,  0.79970997, -0.03009755,  0.34112819,
        -0.15135161,  0.21669497, -0.37680952,  0.04104588],
       [ 0.3844053 ,  0.13920883,  0.07961647,  0.04226906, -0.60610102,
        -0.2751971 , -0.08222008, -0.19654301, -0.57986793],
       [ 0.40990962, -0.37181203, -0.01947537, -0.59412763,  0.15005752,
         0.13497502, -0.5348756 , -0.11252206,  0.02956935],
       [ 0.2591017 ,  0.47413246,  0.12846722, -0.05101188, -0.42012549,
         0.10950357, -0.13990092,  0.04229962,  0.692171  ],
       [ 0.37488902, -0.14148642, -0.14106828, -0.13584327,  0.11883253,
        -0.67856704,  0.3238914 ,  0.43007675,  0.19141608],
       [ 0.27432537, -0.45235526, -0.24105584,  0.1101242 , -0.21117489,
         0.26209582,  0.52658273, -0.45694297,  0.22474374],
       [ 0.47384706, -0.1044102 ,  0.01102628,  0.74672678,  0.2598673 ,
         0.12049856, -0.32105707,  0.1468813 ,  0.01193024],
       [ 0.35341176,  0.29194243,  0.04181639, -0.22636544,  0

In [150]:
# The first eigenvector corresponding to eigenvalue of 3.297793. 
# The following is an array of the vector components.
# Note that the vectors are normalised so that their magnitude is 1.0
eigenvectors[:,0]

array([0.15794137, 0.3844053 , 0.40990962, 0.2591017 , 0.37488902,
       0.27432537, 0.47384706, 0.35341176, 0.16401345])

In [151]:
# The second eigenvector corresponding to eigenvalue of 1.21356187. 
# The following is an array of the vector components.
eigenvectors[:,1]

array([ 0.06862938,  0.13920883, -0.37181203,  0.47413246, -0.14148642,
       -0.45235526, -0.1044102 ,  0.29194243,  0.54045312])

#### (iii) expressions of the first principal component, PC1, and the second principal component, PC2, in terms of the standardized log base 10 values.

#### The principal components can now be expressed in terms of the eigenvector components.


<b>PC1</b> = 0.15794137 x <b>Climate</b> +  0.3844053 x <b>Housing</b> + 0.40990962 x <b>HealthCare</b> + 0.2591017 x <b>Crime</b> + 0.37488902 x <b>Trans</b> +  0.27432537 x <b>Educ</b> + 0.47384706 + <b>Arts</b> + 0.35341176 + <b>Recreat</b> + 0.16401345 x <b>Econ</b>


#### It is to be noted that the data values Climate, Housing, HealthCare and so on are NOT the raw data values but the standardised log tranformed data values.


Similarly,


<b>PC2</b> = 0.06862938 x <b>Climate</b> +  0.13920883 x <b>Housing</b> - 0.37181203 x <b>HealthCare</b> + 0.47413246 x <b>Crime</b> - 0.14148642 x <b>Trans</b> - 0.45235526 x <b>Educ</b> - 0.1044102 x <b>Arts</b> + 0.29194243 x <b>Recreat</b> + 0.54045312 x <b>Econ</b>

#### (iv) the percentage of variation explained by PC1

In [152]:
percent_var_P1 = (eigenvalues[0] / sum(eigenvalues)) * 100
percent_var_P1

36.642144444163485

So PC1 explains 36.64 % of the variation.

In [155]:
# total percentage of variation explained by PC1 PC2.
total_percent_var = (sum(eigenvalues[0:2]) / sum(eigenvalues)) * 100
total_percent_var

50.12616527671426

So PC1 and PC2 explains 50.12 % of the variation.

#### (v) briefly interpret PC1 and PC2.

PC1 = The PC1 expression shows relatively higher positive values of eigenvector components for Healthcare and Arts data. So, PC1 is associated with high Healthcare and Arts quality ratings for a city. If the dimensionality of the actual dataset had to be reduced, Healthcare and Arts could be replaced with PC1 since it is a measure of the quality of healthcare and arts in a city due to the aforementioned reasons.

PC2 = The PC2 expression shows relatively higher eigenvector component values for crime (positive) and education (negative). Thus, it can be said that PC2 represents and distinguishes cities with high crime rates and low education ratings and could be indirectly taken as a measure of crime severity and education qualities of cities.